File: android/src/main/java/expo/modules/mocklivephoto/PlaybackState.kt
package expo.modules.mocklivephoto

internal sealed interface PlaybackPhase {
  data object Idle : PlaybackPhase
  data object Ready : PlaybackPhase
  data object Playing : PlaybackPhase
  data object Paused : PlaybackPhase
  data object Ended : PlaybackPhase
  data object Failed : PlaybackPhase
}

internal class PlaybackState {
  sealed interface Event {
    data object Reset : Event
    data class Ready(val version: Int) : Event
    data class Started(val version: Int) : Event
    data object Pause : Event
    data class Ended(val version: Int) : Event
    data class Failed(val version: Int) : Event
  }

  var phase: PlaybackPhase = PlaybackPhase.Idle
    private set
  var version: Int = 0
    private set

  private var nextReplayToken = 0
  private var pendingReplayToken: Int? = null

  fun reduce(event: Event) {
    when (event) {
      Event.Reset -> {
        version += 1
        phase = PlaybackPhase.Idle
        pendingReplayToken = null
      }
      is Event.Ready -> if (event.version == version && phase == PlaybackPhase.Idle) {
        phase = PlaybackPhase.Ready
      }
      is Event.Started -> if (event.version == version && when (phase) {
        PlaybackPhase.Ready, PlaybackPhase.Paused, PlaybackPhase.Ended -> true
        else -> false
      }) {
        phase = PlaybackPhase.Playing
        pendingReplayToken = null
      }
      Event.Pause -> {
        pendingReplayToken = null
        if (phase == PlaybackPhase.Playing) phase = PlaybackPhase.Paused
      }
      is Event.Ended -> if (event.version == version && phase != PlaybackPhase.Failed) {
        phase = PlaybackPhase.Ended
        pendingReplayToken = null
      }
      is Event.Failed -> if (event.version == version) {
        phase = PlaybackPhase.Failed
        pendingReplayToken = null
      }
    }
  }

  fun requestReplay(version: Int): Int? {
    if (version != this.version || phase != PlaybackPhase.Ended || pendingReplayToken != null) return null
    return (++nextReplayToken).also { pendingReplayToken = it }
  }

  fun consumeReplay(version: Int, token: Int): Boolean {
    if (version != this.version || phase != PlaybackPhase.Ended || pendingReplayToken != token) return false
    pendingReplayToken = null
    return true
  }
}

File: ios/PlaybackState.swift
enum PlaybackPhase: Equatable {
  case idle
  case ready
  case playing
  case paused
  case ended
  case failed
}

struct PlaybackState: Equatable {
  enum Event {
    case reset
    case ready(Int)
    case started(Int)
    case pause
    case ended(Int)
    case failed(Int)
  }

  private(set) var phase: PlaybackPhase = .idle
  private(set) var version = 0
  private var nextReplayToken = 0
  private var pendingReplayToken: Int?

  mutating func reduce(_ event: Event) {
    switch event {
    case .reset:
      version += 1
      phase = .idle
      pendingReplayToken = nil
    case .ready(let eventVersion) where eventVersion == version && phase == .idle:
      phase = .ready
    case .started(let eventVersion)
      where eventVersion == version && (phase == .ready || phase == .paused || phase == .ended):
      phase = .playing
      pendingReplayToken = nil
    case .pause:
      pendingReplayToken = nil
      if phase == .playing {
        phase = .paused
      }
    case .ended(let eventVersion) where eventVersion == version && phase != .failed:
      phase = .ended
      pendingReplayToken = nil
    case .failed(let eventVersion) where eventVersion == version:
      phase = .failed
      pendingReplayToken = nil
    default:
      break
    }
  }

  mutating func requestReplay(version eventVersion: Int) -> Int? {
    guard eventVersion == version, phase == .ended, pendingReplayToken == nil else { return nil }
    nextReplayToken += 1
    pendingReplayToken = nextReplayToken
    return nextReplayToken
  }

  mutating func consumeReplay(version eventVersion: Int, token: Int) -> Bool {
    guard eventVersion == version, phase == .ended, pendingReplayToken == token else { return false }
    pendingReplayToken = nil
    return true
  }
}
